### ISSUE TITLE
cmd/go: 'go mod vendor' should not remove existing non-Go content from the vendor subdirectory

### ISSUE THREAD

**Issue by reitzig** [golang/go#44119]
**Tl;dr**: Go is not the only tool I use in my project; I expect it share `vendor` peacefully.

### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.15.8 linux/amd64
</pre>

(latest at the time of this writing)

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/raphael/.cache/go-build"
GOENV="/home/raphael/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/raphael/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/raphael/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/raphael/Code/tools/container-do/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build500284300=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

### What did you do?

```Gemfile
# Gemfile
source "https://rubygems.org"

group :test do
  gem 'cucumber', '~> 5'
  gem 'rspec', '~> 3'
end
```

```
# go.mod
module github.com/reitzig/container-do

go 1.15

require (
	github.com/pelletier/go-toml v1.8.1
	go.uber.org/zap v1.16.0
)
```

```bash
bundle2.7 config set --local path 'vendor/bundle'
bundle2.7 install
go mod vendor
```

### What did you expect to see?

Ruby gems _and_ Go modules "installed" in `vendor`.

### What did you see instead?

Only Go modules in `vendor`.

### Workaround

Bundler leaves Go modules alone, so re-running `bundle install` after `go mod vendor` creates the desired state.

**Comment by mvdan** [comment]
This seems like a duplicate of https://github.com/golang/go/issues/26366.

**Comment by reitzig** [comment]
Very old and unsolved, but yes, essentially the same issue. Don't know if, after all this time, it refers to the same implementation anymore.

**Comment by bcmills** [comment]
Is there a typo in the issue title? I'm having trouble figuring out what the desired behavior is here.

**Comment by reitzig** [comment]
> Is there a typo in the issue title?

I don't think so. What's your understanding of the expected behaviour?

**Comment by bcmills** [comment]
Ok, I think I understand now — this is the same problem as #44646, but **not** the same problem as #26366?

**Comment by reitzig** [comment]
Yes. Seems I misread #26366 on first reading, my bad. (Edited above comment to hopefully make things clearer to future readers.)

**Comment by fgergo** [comment]
@bcmills why is WaitingForInfo label applied?
"go mod vendor" nukes a complete directory without any warning. Please share what info do you think is needed to progress with this issue?
If a different expected behaviour, then as a first approximation I believe this should suffice:
not deleting a folder without explicit confirmation

**Comment by bcmills** [comment]
@fgergo, my apologies. I added the WaitingForInfo tag when trying to determine what exactly the issue was that was being reported, and forgot to remove it when the issue was clarified.

**Comment by fgergo** [comment]
@bcmills @seankhliao I'm not sure current issue title is precise: "cmd/go: 'go mod vendor' should not remove existing non-Go content from the vendor subdirectory"
"go mod vendor" removes everything when it comes to that conclusion without any warning. It removes both go code and other files as well like I described in #45355 .
I believe a title like
"cmd/go: 'go mod vendor' should not remove existing content from the vendor subdirectory without any warning"
describes the situation more precisely.

**Comment by reitzig** [comment]
To be clear: _my_ intent was for `go mod` to leave all files it isn't responsible for in peace. A warning would be good either way, but _I_ don't want it to remove the Ruby files at all! It can do whatever it wants with _Go_ files for all I care since that's its job.

If it insists on completely owning that directory, it needs to be configurable.

**Comment by esimov** [comment]
Are there any updates related to this issue? It seems that there are no activities going on recently addressing this problem. This issue is still happening with go1.18.3.

**Comment by lordofscripts** [comment]
I agree and don't understand why the prvious one was closed without FIX.

I had just moved two external single-directory libraries (each with 2-5 GO files) into the `vendor` directory because I wanted to stress that those two were not written by me (_just modified_).

Then after moving the two directories into `vendor/`  I ran the `go mod vendor` and when I went to re-compile PLOP! the `go mod vendor` command had **deleted those two directories under vendor without asking!**

Now, all my modifications were gone! luckily a few days back I had tar-balled my application but still, I lost several important changes. The `go mod vendor` command you _at least warn you what it is going to do_ and ask for confirmation, not just delete without asking. Yes, I consider it a bug.

**Comment by RazvanML** [comment]
Simply having hardcoded "special" names like vendor seems plain wrong.

**Comment by ianlancetaylor** [comment]
@RazvanML That decision was made long ago and it is not going to change. There are other special names as well, such as "testdata"; see https://pkg.go.dev/cmd/go#hdr-Package_lists_and_patterns.

**Comment by bcsgh** [comment]
While this is being talked about; there is also an argument for providing a way to run `go mod vendor` where it will never delete anything _including_ things it "owns". The specific provoking cases I'm seeing is where multiple concurrent invocations are stomping on each other (the long term fix is to avoid that situation, but external "non technical" constraints make that not practicable in the needed timeframe and go needs to be able to deal with other than ideal environments of which my situation is one of many).

FWIW, in such a mode, if a invocation can't produce a correct result by adding new files (e.g. a required file already exist but with the wrong content) an error would be preferable to overwriting the file.

**Comment by matloob** [comment]
At this point, it's unlikely for us to change this behavior. go expects to own the vendor directory in a module. The implementation would also have some complexity: while vendor/module.txt does list the packages that are vendored in the vendor directory, we also vendor embedded files, which means that we can have non-go files in the vendor directory that are managed by go.
